recommenders/models/ncf/dataset.py

```diff
@@ -1,321 +1,450 @@
+                col_user=self.col_user,
+                col_item=self.col_item,
+                col_rating=self.col_rating,
+                col_test_batch=self.col_test_batch,
+                binary=self.binary
             )
-            self.id2item = {self.item2id[k]: k for k in self.item2id}
-
-        return self._reindex(train, binary), self._reindex(test, binary)
-
-    def _reindex(self, df, binary):
-        """Process dataset to reindex userID and itemID, also set rating as binary feedback
-
-        Args:
-            df (pandas.DataFrame): dataframe with at least columns (col_user, col_item, col_rating)
-            binary (bool): if true, set rating>0 to rating = 1
 
-        Returns:
-            list: train and test pandas.DataFrame Dataset, which have been reindexed.
-
-        """
-
-        # If testing dataset is None
-        if df is None:
-            return None
-
-        # Map user_idx and item_idx
-        df = pd.merge(df, self.user_idx, on=self.col_user, how="left")
-        df = pd.merge(df, self.item_idx, on=self.col_item, how="left")
-
-        # If binary feedback, set rating as 1.0 or 0.0
-        if binary:
-            df[self.col_rating] = df[self.col_rating].apply(lambda x: float(x > 0))
+        # set random seed
+        random.seed(seed)
 
-        # Select relevant columns
-        df_reindex = df[
-            [self.col_user + "_idx", self.col_item + "_idx", self.col_rating]
+    def _sample_negatives_with_replacement(self, user_negative_item_pool, n_samples):
+        return random.choices(user_negative_item_pool, k=n_samples)
+
+    def _sample_negatives_without_replacement(self, user_negative_item_pool, n_samples):
+        return random.sample(user_negative_item_pool, k=n_samples)
+
+    def _get_user_negatives_pool(self, user_positive_item_pool):
+        # get list of items user has not interacted with
+        return list(set(self.train_datafile.items) - user_positive_item_pool)
+
+    def _get_negative_examples(self, user, user_negative_item_pool, n_samples):
+        # create dataframe containing negative examples for user assigned zero rating
+        user_negative_samples = self._sample(user_negative_item_pool, n_samples)
+        return pd.DataFrame({
+            self.col_user: [user] * n_samples,
+            self.col_item: user_negative_samples,
+            self.col_rating: [0.0] * n_samples
+        })
+
+    def _check_sample_size(self, user, n_samples, user_negative_item_pool, training=True):
+        # if sampling without replacement, check sample population is sufficient and reduce
+        # n_samples if not.
+        n_neg_var = "n_neg" if training else "n_neg_test"
+        dataset_name = "training" if training else "test"
+
+        new_n_samples = n_samples
+        if not self.sample_with_replacement:
+            population_size = len(user_negative_item_pool)
+            if n_samples > population_size:
+                new_n_samples = min(n_samples, population_size)
```

Comment:
How about incorporating this as `k=min(n_samples, len(user_negative_item_pool))` inside `_sample_negatives_without_replacement()` and do away with this function?
Another question: how is the negative sampling according to the paper, with a fixed sample size or proportional to the user's population size?

Reply:
It's possible. The reason I do this outside of `_sample_negatives_without_replacement` is that in `_create_test_file` I have to call `_sample_negatives_without_replacement` for every positive example for a user. If the sample size needs to be adjusted I raise a warning and ideally I'd only want to do this once per user rather than for every positive example. However, I should be able to parametrize `_sample_negatives_without_replacement` such that it would only raise a warning the first time it is called for each user. I'll make this change.

In the paper, 4 negative examples are sampled for every positive example so it's the ratio of positive to negative examples that are fixed for each user. A user with more positive examples will therefore have more negative examples. In the original code, this sampling is achieved by calling the sampling function once per positive example. However, I changed this because it's less efficient but also because it effectively amounts to sampling _with_ replacement as the same negative sample can appear more than once per user. In my code (during training and not for test file creation), I perform the sampling once per user without replacement so that the negatives for a user are unique. This also means however that the number of requested samples sometimes exceeds the population of negative items for a user and so some users have a smaller ratio of positives to negatives. I'm not sure which approach is better. The choices are:

1. Sample without replacement per user, guaranteeing complete diversity of negative samples, but having a reduced ratio of positive/negative samples for some users [my current code]
2. Sample without replacement per positive example, ensuring the ratio of positive/negative samples are the same for all users, but some negative examples will appear twice for a given user [old code]
3. Sample _with_ replacement per user, which is the same as option 2 but more efficient and guarantees equal ratios across users.
4. Sample without replacement per user, and if the resulting number of negative samples is less than required, sample again to make up the shortfall. This would be efficient and the keep the ratios constant across users but some negative samples will appear more than once (but only for users that have limited population sizes).
Do you have any thoughts on what is the best approach?

Reply:
It looks like they do #2 in [their code](https://github.com/hexiangnan/neural_collaborative_filtering/blob/4aab159e81c44b062c091bdaed0ab54ac632371f/GMF.py#L92) but with a fixed num_negatives (not a proportion). I am not sure how they get their test data, it is available as a raw file. The paper says there are 100 test samples for each user.
I think with the data sets they use it is feasible to sample without duplicates (since there are many items and large sparsity).
I agree that duplicate examples are not a good idea, for reasons of efficiency but also because they effectively give more importance to some training examples compared to others. So I think your approach is good. The problems arise only when the sparsity is low (at least for some users), but then it's a matter for the user to decide what to do as a pre-processing step (it depends on each specific case).

Reply:
@anargyri  I've tried moving the logic of `_check_sample_size` into the `_sample_negatives_without_replacement` function but the result is quite messy. If we want to just reduce the number of samples as you suggest it is no problem but if we also want to _warn_ users about it then I have to pass a bunch of extra information to the sampling function (as well as the `_get_negative_examples` function that calls it) in order to make the warnings meaningful. Plus I have to add an additional if statement to ensure we only print warnings once per user when creating the test set. I'm not sure it's improved anything by doing this but I can make the change if you want. What do you think?

Reply:
Ok then, better keep it as is.